lib/tapioca/dsl/compiler.rb

```diff
@@ -25,6 +25,8 @@ class Compiler
       sig { returns(T::Hash[String, T.untyped]) }
       attr_reader :options
 
+      @@requested_constants = T.let([], T::Array[Module])
```

Comment:
This sketches me out a little, but there's a really really nice benefit to this approach: all the existing DSL Compilers will mostly just work, without needing to be updated.

What if we loosened that constraint: could we make a better API if we were willing to change the compilers?

e.g. what if the main module selection was done by something like:

```ruby
sig { abstract.params(candidates: T::Enumerable[Module]).returns(T::Enumerable[Module]) }
def select_constants_to_decorate(candidates)
```

This would replace `gather_constants`. By default, we would call it with `select_constants_to_decorate(all_modules)` (which would work similar to how gather_constants works in practice today), otherwise we could do `select_constants_to_decorate(requested_constants)`.